subsys/bluetooth/audio/bap_broadcast_assistant.c

```diff
@@ -184,15 +192,148 @@ static int parse_recv_state(const void *data, uint16_t length,
 	return 0;
 }
 
+static uint8_t parse_and_send_recv_state(struct bt_conn *conn, const void *data,
+					 uint16_t length,
+					 struct bt_bap_scan_delegator_recv_state *recv_state)
+{
+	int err;
+	int16_t index;
+
+	err = parse_recv_state(data, length, recv_state);
+	index = lookup_index_by_handle(broadcast_assistant.handle);
+
+	if (err != 0) {
+		LOG_WRN("Invalid receive state received");
+		if (conn == NULL) {
+			LOG_DBG("Unexpected conn");
+		} else {
+			LOG_DBG("conn ok");
+		}
+		return BT_GATT_ITER_STOP;
+	}
+
+	broadcast_assistant.src_ids[index] = recv_state->src_id;
+	broadcast_assistant.past_avail[index] = past_available(conn,
+							       &recv_state->addr,
+							       recv_state->adv_sid);
+
+	if (broadcast_assistant_cbs != NULL &&
+	    broadcast_assistant_cbs->recv_state != NULL) {
+		broadcast_assistant_cbs->recv_state(conn, 0, recv_state);
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t broadcast_assistant_bap_ntf_read_func(struct bt_conn *conn, uint8_t err,
+						     struct bt_gatt_read_params *read,
+						     const void *data, uint16_t length)
+{
+	struct bt_bap_scan_delegator_recv_state recv_state;
+	uint16_t handle = read->single.handle;
+	struct net_buf_simple buf_clone;
+	struct net_buf_simple *buf;
+
+	LOG_DBG("conn %p err 0x%02x len %u", conn, err, length);
+
+	if (err) {
+		LOG_DBG("Failed to read: %u", err);
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	LOG_DBG("handle 0x%04x", handle);
+
+	buf = &broadcast_assistant.net_buf;
+
+	if (data != NULL) {
+		if (net_buf_simple_tailroom(buf) < length) {
+			LOG_DBG("Buffer full, invalid server response of size %u",
+				length + buf->len);
+			broadcast_assistant.busy = false;
+			net_buf_simple_reset(buf);
+
+			return BT_GATT_ITER_STOP;
+		}
+
+		broadcast_assistant.busy = false;
+		/* store data*/
+		net_buf_simple_add_mem(buf, data, length);
+		return parse_and_send_recv_state(conn, buf, buf->len, &recv_state);
```

Comment:
If `parse_and_send_recv_state` fails, then we are also missing a reset of the buffer. 

Effectively the buffer needs to be reset on read complete or read error (whenever the read procedure ends). 

Ideally we should not call `parse_and_send_recv_state` until after we've read everything; else we are potentially sending partial receive states to the application.

Reply:
So bt_gatt_read can call us before it has received all the data? How  can we know that all the data has been received?

Reply:
> So bt_gatt_read can call us before it has received all the data? How can we know that all the data has been received?

Yes, for long reads, the read callback is called for each segment. You know that all data has been read when `err == 0` and `data == NULL`